File: algorithms/sort/quick-sort/quicksort.c
#include <stdio.h>

void swap (int *a, int *b);

void quicksort (int *arr, int left, int right) {
  if (left >= right) {
    return;
  }

  int pivot = arr[left];
  int i = left + 1, j = right;

  while (1) {
    while (i <= right) {
      if (arr[i] > pivot) {
        break;
      }
      i++;
    }

    while (j > left) {
      if (arr[j] < pivot) {
        break;
      }
      j--;
    }

    if (i > j) {
      break;
    }
    swap(&arr[i], &arr[j]);
  }

  swap(&arr[left], &arr[j]);

  quicksort(arr, left, j - 1);
  quicksort(arr, j + 1, right);
}

void swap (int *a, int *b) {
  int tmp = *a;
  *a = *b;
  *b = tmp;
}

File: algorithms/sort/quick-sort/quick-sort.rb
def quick_sort arr, left, right
  return arr if left >= right

  pivot = arr[left]
  i = left + 1
  j = right

  while true
    while i <= right
      break if arr[i] > pivot
      i += 1
    end

    while j > left
      break if arr[j] < pivot
      j -= 1
    end

    break if i > j
    arr[i], arr[j] = arr[j], arr[i]
  end

  arr[left], arr[j] = arr[j], arr[left]

  quick_sort arr, left, j - 1
  quick_sort arr, j + 1, right

  arr
end
